src/ring_theory/localization.lean

```diff
@@ -436,10 +436,17 @@ ring_hom.ext $ monoid_hom.ext_iff.1 $ (to_localization_map M S).lift_of_comp j.t
 
 variables (M)
 
+/-- To show `j` and `k` agree on the whole localization, it suffices to show they agree
+on the image of the base ring, if they preserve `1` and `*`. -/
+protected lemma ext (j k : S → P) (hj1 : j 1 = 1) (hk1 : k 1 = 1)
```

Comment:
Why not make `j` and `k` be monoid_homs? Then you can tag this @[ext]

Reply:
Because that would be exactly `submonoid.localization_map.epic_of_localization_map` (or `is_localization.epic_of_localization_map` if they are `ring_hom`s). [In my intended application](https://github.com/leanprover-community/mathlib/pull/8701/files#diff-52f399748c2df16138ee6af28a862c2bc4a71efc7f0d84436c3d1c8719e8caf2R61), `j` and `k` are the composition of an `alg_equiv` and a `ring_hom` resp. the composition of an `alg_equiv` and `smul`, so keeping them unbundled is way more practical.

Reply:
What does `epic` mean in this context?

Reply:
These are the two lemmas I'd recommend:
```
/-- See note [partially-applied ext lemmas] -/
@[ext]
lemma monoid_hom_ext ⦃j k : S →* P⦄
  (h : j.comp (algebra_map R S : R →* S) = k.comp (algebra_map R S)) : j = k :=
submonoid.localization_map.epic_of_localization_map (to_localization_map M S) $
  monoid_hom.congr_fun h

/-- See note [partially-applied ext lemmas] -/
@[ext]
lemma ring_hom_ext ⦃j k : S →+* P⦄
  (h : j.comp (algebra_map R S) = k.comp (algebra_map R S)) : j = k :=
ring_hom.coe_monoid_hom_injective $ monoid_hom_ext M $ monoid_hom.ext $ ring_hom.congr_fun h
```
Feel free to keep yours as well, but these ones make the `ext` tactic stronger.

Reply:
> What does `epic` mean in this context?

[It's the adjective form of epimorphism, from category theory](https://en.wikipedia.org/wiki/Epimorphism). That is, `algebra_map R S` is epic because composing on the right with that map is injective.

Reply:
Ah, apologies; I now see these are not elegible for use as `ext` lemmas as `M` is not determined.